### PR TITLE
Retry commits that failed during a rebalance

### PR DESCRIPTION
**Pull request by altomch**
References #1243

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala [hunk 1]
```diff
@@ -240,12 +241,6 @@ import scala.util.control.NonFatal
   private var commitRefreshing: CommitRefreshing = _
   private var stopInProgress = false
 
-  /**
-   * While `true`, committing is delayed.
-   * Changed by `onPartitionsRevoked` and `onPartitionsAssigned` in [[RebalanceListenerImpl]].
-   */
-  private var rebalanceInProgress = false
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala:247]: Nice to get rid of this :+1:

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala [hunk 2]
```diff
@@ -608,14 +603,19 @@ import scala.util.control.NonFatal
               commitRefreshing.committed(offsets)
               replyTo.foreach(_ ! Done)
 
+            case e: RebalanceInProgressException =>
+              log.info("Kafka commit is to be retried, after={} ms, commitsInProgress={}, cause={}",
+                       duration / 1000000L,
+                       commitsInProgress,
+                       e)
+              postponeCommits(commitMap, replyTo)
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala:611]: `retryCommits` sounds better to me.

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala [hunk 3]
```diff
@@ -608,14 +603,19 @@ import scala.util.control.NonFatal
               commitRefreshing.committed(offsets)
               replyTo.foreach(_ ! Done)
 
+            case e: RebalanceInProgressException =>
+              log.info("Kafka commit is to be retried, after={} ms, commitsInProgress={}, cause={}",
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala:607]: Include the logging in the method too to reduce duplication.

File: tests/src/test/scala/akka/kafka/scaladsl/CommittingSpec.scala [hunk 4]
```diff
@@ -264,6 +264,118 @@ class CommittingSpec extends SpecBase with TestcontainersKafkaLike with Inside {
       control2.isShutdown.futureValue shouldBe Done
     }
 
+    "retry commits that occurred during rebalance" in assertAllStagesStopped {
```
Comment (seglo) [inline tests/src/test/scala/akka/kafka/scaladsl/CommittingSpec.scala:267]: This test is quite complex.  See review summary comment.

**Review by seglo — COMMENTED** [review]
@altomch I accidentally worked on a solution to your reported issue before I realized you created this PR, sorry.  I closed my PR, but I wonder if yours could be made simpler.  I'm worried the test in particular will be a challenge to maintain in the future.  If you look at the test I created in my closed PR I wrote a unit test in `CommittingWithMockSpec` that doesn't require an actual Kafka broker. I think this would be the preferable way to go because it's easier to understand and fits the convention of the existing commit retry tests. Would you mind adopting that test instead?

https://github.com/akka/alpakka-kafka/pull/1247

BTW, it looks like you need to sign the CLA. That will need to be done before this PR can be merged.

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala (hunk 2 above)
Comment (altomch) [reply core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala]: agree, thanks

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala (hunk 3 above)
Comment (altomch) [reply core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala]: I thought that we don't need "warning" level on such retries, but I don't mind

File: tests/src/test/scala/akka/kafka/scaladsl/CommittingSpec.scala (hunk 4 above)
Comment (altomch) [reply tests/src/test/scala/akka/kafka/scaladsl/CommittingSpec.scala]: thank you, removed this test in favour of unit one
the only reason to add this test was removing of ```rebalanceInProgress``` flag, but I've realised that it had been redundant before my changes.

**Comment by altomch** [comment]
@seglo, thank you for review
I've already solved issue with CLA, at least check is passed.

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala [hunk 5]
```diff
@@ -608,14 +613,8 @@ import scala.util.control.NonFatal
...
-            case e: RetriableCommitFailedException =>
-              log.warning("Kafka commit is to be retried, after={} ms, commitsInProgress={}, cause={}",
-                          duration / 1000000L,
-                          commitsInProgress,
-                          e.getCause)
-              commitMaps = commitMap.toList ++ commitMaps
-              commitSenders = commitSenders ++ replyTo
-              requestDelayedPoll()
+            case e: RebalanceInProgressException => retryCommits(duration, e)
+            case e: RetriableCommitFailedException => retryCommits(duration, e)
```
Comment (seglo) [inline core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala:617]:
```suggestion
            case e: RetriableCommitFailedException => retryCommits(duration, e.getCause)
```

**Review by seglo — COMMENTED** [review]
Looking good, just one fix.

File: core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala (hunk 5 above)
Comment (altomch) [reply core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala]: thanks, overlooked

**Comment by seglo** [comment]
Unrelated build failure

**Review by seglo — APPROVED** [review]
LGTM

**Pull request merged**